---
title: "Clustering and categorization"
author: "Peter Ralph"
date: "26 February 2018 -- Advanced Biological Statistics"
---

```{r setup, include=FALSE}
fig.dim <- 4
knitr::opts_chunk$set(fig.width=2*fig.dim,
                      fig.height=fig.dim,
                      fig.align='center')
set.seed(23)
library(lars)
library(tidyverse)
library(rstan)
library(matrixStats)
options(mc.cores = parallel::detectCores())
```

# Clusters in expression space

## A conceptual model

Let's build a *conceptual* model for
descriptive analysis of "mixture" expression data.

. . .

**Data:** expression data from tissue samples
that consist of various *mixtures* of different cell types.

. . .

**Goal:** identify shared coexpression patterns
corresponding to *cell type*.

. . .

*Similar situations:* 
identify different developmental stages from whole-organism expression;
common community structures from metagenomic data.

----------------


1. Each cell type has a typical set of *mean* expression levels.

2. Each sample is composed of a mixture of cell types,
   defined by the proportions that come from each type.

3. Mean expression levels differ between cell types
   for only some of the genes.

4. Some samples are *noisier* than others.

. . .

**Assume** the same amount of sequencing of each sample.
   

------------------

::: {.columns}
:::::::::::::: {.column width=50%}


1. Mean expression by cell type.

2. Cell type proportions by sample.


:::
:::::::::::::: {.column width=50%}


1. $x_{kj}$ : Mean expression of gene $j$ in cell type $k$.

2. $w_{ik}$ : Proportion of sample $i$ of cell type $k$.

$Z_{ij}$ : expression level in sample $i$ of gene $j$.

   $$\begin{aligned}
        Z_{ij} \approx \sum_{k=1}^K w_{ik} x_{kj} .
   \end{aligned}$$


:::
:::::::::::::: 

------------------

::: {.columns}
:::::::::::::: {.column width=50%}


1. Mean expression by cell type.

2. Cell type proportions by sample.

3. Mean expression levels differ between cell types
   for only some of the genes.

4. Some samples are *noisier* than others.

:::
:::::::::::::: {.column width=50%}


$Z_{ij}$ : expression level in sample $i$ of gene $j$.

   $$\begin{aligned}
        Z_{ij} \approx \sum_{k=1}^K w_{ik} x_{kj} .
   \end{aligned}$$

3. $y_j$, $\eta_j$ : mean and SD of expression of gene $j$ across *all* cell types;
   shrink $x_{kj}$ towards $y_j$.

4. *(omit this)*

:::
:::::::::::::: 

----------------------

::: {.columns}
:::::::::::::: {.column width=50%}

```
```{r nmf1, echo=FALSE, results="asis", cache=TRUE}
cat(nmf1 <- "data {
  int N; // # samples
  int L; // # genes
  int K; // # cell types
  int Z[N,L];
}")
```
```
```
```{r nmf2, echo=FALSE, results="asis", cache=TRUE}
cat(nmf2 <- "parameters {
  matrix<lower=0>[L,K] x;
  vector[L] y;
  simplex[K] w[N];
  vector<lower=0>[L] eta;
  vector<lower=0>[K] alpha;
  real<lower=0> d_alpha;
}")
```
```
```
```{r nmf3, echo=FALSE, results="asis", cache=TRUE}
cat(nmf3 <- "model {
  for (i in 1:N) {
      Z[i] ~ poisson(eta .* (x * w[i]));
      w[i] ~ dirichlet(d_alpha * alpha);
  }
  for (j in 1:K) 
      { x[,j] ~ normal(y ./ eta, 1); }
  y ~ normal(0, 20);
  alpha ~ normal(0, 1);
  d_alpha ~ exponential(0.2);
  eta ~ cauchy(0, 10);
}")
```
```


:::
:::::::::::::: {.column width=50%}


1. $x_{kj}$ : Mean expression of gene $j$ in cell type $k$.

2. $w_{ik}$ : Proportion of sample $i$ of cell type $k$.

$$\begin{aligned}
    Z_{ij} \approx \sum_k w_{ik} x_{kj} .
\end{aligned}$$

3. $y_j$, $\eta_j$ : mean and SD of expression of gene $j$ across *all* cell types;
   shrink $x_{kj}$ towards $y_j$.


:::
:::::::::::::: 

## Testing: compiles?

```{r model_compiles, echo=-1, cache=TRUE, dependson=c("nmf1", "nmf2", "nmf3")}
nmf_block <- paste(nmf1, nmf2, nmf3, sep='\n')
nmf_model <- stan_model(model_code=nmf_block)
```

## Testing: runs?

```{r test_runs}
sampling(nmf_model,
         data=list(N=10,
                   L=5,
                   K=2,
                   Z=matrix(rpois(50, 100), ncol=5)),
         chains=1, iter=100)
```


# Nonnegative matrix factorization

## ... aka "NMF"


We are *decomposing* $Z$ into the product of two lower-dimensional,
nonnegative factors:

$$\begin{aligned}
    Z_{ij} &\approx \sum_k w_{ik} x_{kj} \\
    w_{ik} &\ge 0 \\
    x_{kj} &\ge 0 .
\end{aligned}$$

This is like PCA, but sometimes more *interpretable*.


# Stochastic minute

## the Dirichlet distribution

A random set of $k$ *proportions* $0 \le P_i \le 1$
has a $\Dirichlet(\alpha_1, \ldots, \alpha_k)$ if it has probability density
$$\begin{aligned}
    \frac{1}{B(\alpha)} \prod_{i=1}^k p_i^{\alpha_i} 
\end{aligned}$$
over the set of possible values
$$\begin{aligned}
    P_1 + \cdots + P_k = 1 .
\end{aligned}$$

. . .

1. This is useful as a prior on *proportions*.

2. This generalized the Beta: if $X \sim \Beta(a, b)$ then $(X, 1-X) \sim \Dirichlet(a, b)$.

3. Marginal distributions are Beta distributed: $P_i \sim \Beta(\alpha_i, \sum_{j=1}^k \alpha_j - \alpha_i)$.

-----------------

4. If $X_i \sim \Exp(\alpha_i)$, and
   $$\begin{aligned}
    P_i = X_i / \sum_{j=1}^k X_j
   \end{aligned}$$
   then $P \sim \Dirichlet(\alpha)$.

## "Simplex" parameters

"The $k$-simplex" is the set of *proportions*,
i.e., nonnegative numbers $p$ satisfying
$$\begin{aligned}
    p_1 + \cdots p_k = 1 .
\end{aligned}$$

```
parameters {
    simplex[K] w[N];
}
model {
    w ~ dirichlet(d_alpha * alpha);
}
```

# Simulate data

## Outline

1. How many cell types? *five*

2. How many genes? *one thousand*

3. How many samples? *one hundred*

4. How much noise in expression?
   *Mean expression per gene are $\mu \sim \Exp(1/100)$;
    expression per sample is $\Poisson(m)$,
    where $m$ is truncated $\Normal(\mu, \sigma \mu)$,
    and $\sigma$ is around 1.*

5. How many genes distinguish cell types,
   and by how much relative to expression?
   *Ten percent of genes, and a gene with mean expression $\mu$
   differs between cell types by about $\mu/2$:
   means by cell type are $\mu \times \log\Normal(0, 1)$.*

6. How many "noisy" genes?  How many "noisy" samples?
   *Two percent of each.*


# An easy case

## Parameters:

```{r sim_data, cache=TRUE}
ntypes <- 3
ngenes <- 1000
nsamples <- 1000
ninf <- 0.6 * ngenes # number of informative genes
```

-----------

Determine expression profiles per cell type
```{r sim_data_2, cache=TRUE, dependson="sim_data"}
mu <- rexp(ngenes, .01)
# copy mu into five columns
type_mu <- do.call(cbind, list(mu)[rep(1,ntypes)])

# simulate how differnt mu is across cell types
diff_sigma <- matrix(rlnorm(ntypes * ninf, log(0.5), 1), ncol=ntypes)
type_mu[1:ninf,] <- type_mu[1:ninf,] * diff_sigma

m_list <- lapply(1:ntypes, function (ct) {
                    matrix(pmax(0, rnorm(ngenes*nsamples, 
                                         type_mu[,ct], type_mu[,ct])),
                           ncol=ngenes, byrow=TRUE) })
```

---------

Simulate proportions per sample, and construct expected expression levels
```{r sim_data_3, cache=TRUE, dependson="sim_data_2"}
# simulate proportions per sample
w <- matrix( rexp(nsamples*ntypes), ncol=ntypes)
w <- sweep(w, 1, rowSums(w), "/")
stopifnot(all(abs(rowSums(w) - 1) < 1e-15))

# construct expected expression levels:
m <- matrix(0, nrow=nsamples, ncol=ngenes)
for (ct in 1:ntypes) {
    m <- m + (w[,ct] * m_list[[ct]])
}
```

-------------

```{r sim_data_4, cache=TRUE, dependson="sim_data_3"}
# simulate expression
Z <- matrix(rpois(length(m), m), ncol=ncol(m))

# add noisy indivs, genes??

```

## Point estimates with `optimizing`

```{r optim_nmf, cache=TRUE, dependson=c("model_compiles", "sim_data_4")}
system.time(nmf_optim <- optimizing(nmf_model,
                                    data=list(N=nsamples,
                                              L=ngenes,
                                              K=ntypes,
                                              Z=Z)))
```

## Can we infer mixture proportions?

```{r optim_results}
results <- list(x=nmf_optim$par[grepl("^x", names(nmf_optim$par))],
                y=nmf_optim$par[grepl("^y", names(nmf_optim$par))],
                w=nmf_optim$par[grepl("^w", names(nmf_optim$par))],
                eta=nmf_optim$par[grepl("^eta", names(nmf_optim$par))],
                alpha=nmf_optim$par[grepl("^alpha", names(nmf_optim$par))],
                d_alpha=nmf_optim$par[grepl("^d_alpha", names(nmf_optim$par))])
dim(results$x) <- c(ngenes, ntypes)
dim(results$w) <- c(nsamples, ntypes)

cor(results$w, w)
```

------------------

```{r show_optim, echo=FALSE, fig.width=3*fig.dim, fig.height=2*fig.dim}
w_pos <- svd(w)$u[,1]
w_cols <- rainbow(64)[ceiling(64*(w_pos - min(w_pos))/diff(range(w_pos)))]
layout(matrix(1:ntypes^2, nrow=ntypes))
for (i in 1:ntypes) {
    for (j in 1:ntypes) {
        plot(w[,i], results$w[,j], col=w_cols, pch=20,
             xlab="true w", ylab="inferred w", main=paste(i, j))
        abline(0,1)
    }
}
```

------------------

```{r match_w, echo=FALSE}
wij <- apply(cor(results$w, w), 2, which.max)
mean_abs_err <- mean(abs(w - results$w[,wij]))
```

The mean absolute error in inference of mixture proportions ($w$)
is `r mean_abs_err`.


## Can we infer expression profiles?

```{r x_results}
cor(results$x, type_mu)
```

----------------


```{r x_plot_res, fig.width=4*fig.dim}
xij <- apply(cor(results$x, type_mu), 2, which.max)
x_ord <- apply(type_mu, 2, order)
layout(t(1:ntypes))
for (i in 1:ntypes) {
    plot(type_mu[x_ord[,i],i], pch=20,
         main=paste("expression profile", i))
    points((results$eta * results$x[,xij[i]])[x_ord[,i]], pch=20, col='red')
}
legend("topleft", pch=20, col=c("black", "red"),
       legend=c("truth", "inferred"))
```

# A harder case

## Parameters:
```{r hsim_data, cache=TRUE}
ntypes <- 5
ngenes <- 1000
nsamples <- 100
ninf <- 0.1 * ngenes # number of informative genes
```

---------

Determine expression profiles per cell type
```{r hsim_data_2, cache=TRUE, dependson="hsim_data"}
mu <- rexp(ngenes, .01)
# copy mu into five columns
type_mu <- do.call(cbind, list(mu)[rep(1,ntypes)])

# simulate how differnt mu is across cell types
diff_sigma <- matrix(rlnorm(ntypes * ninf, log(0.5), 1), ncol=ntypes)
type_mu[1:ninf,] <- type_mu[1:ninf,] * diff_sigma

m_list <- lapply(1:ntypes, function (ct) {
                    matrix(pmax(0, rnorm(ngenes*nsamples, 
                                         type_mu[,ct], type_mu[,ct])),
                           ncol=ngenes, byrow=TRUE) })
```

---------

Simulate proportions per sample, and construct expected expression levels
```{r hsim_data_3, cache=TRUE, dependson="hsim_data_2"}
# simulate proportions per sample
w <- matrix( rexp(nsamples*ntypes), ncol=ntypes)
w <- sweep(w, 1, rowSums(w), "/")
stopifnot(all(abs(rowSums(w) - 1) < 1e-15))

# construct expected expression levels:
m <- matrix(0, nrow=nsamples, ncol=ngenes)
for (ct in 1:ntypes) {
    m <- m + (w[,ct] * m_list[[ct]])
}
```

-------------

```{r hsim_data_4, cache=TRUE, dependson="hsim_data_3"}
# simulate expression
Z <- matrix(rpois(length(m), m), ncol=ncol(m))

# add noisy indivs, genes??

```

## Point estimates with `optimizing`

```{r hoptim_nmf, cache=TRUE, dependson=c("model_compiles", "hsim_data_4")}
system.time(nmf_optim <- optimizing(nmf_model,
                                    data=list(N=nsamples,
                                              L=ngenes,
                                              K=ntypes,
                                              Z=Z)))
```

## Can we infer mixture proportions?

```{r hoptim_results}
results <- list(x=nmf_optim$par[grepl("^x", names(nmf_optim$par))],
                y=nmf_optim$par[grepl("^y", names(nmf_optim$par))],
                w=nmf_optim$par[grepl("^w", names(nmf_optim$par))],
                eta=nmf_optim$par[grepl("^eta", names(nmf_optim$par))],
                alpha=nmf_optim$par[grepl("^alpha", names(nmf_optim$par))],
                d_alpha=nmf_optim$par[grepl("^d_alpha", names(nmf_optim$par))])
dim(results$x) <- c(ngenes, ntypes)
dim(results$w) <- c(nsamples, ntypes)

cor(results$w, w)
```

------------------

```{r hshow_optim, echo=FALSE, fig.width=3*fig.dim, fig.height=2*fig.dim}
w_pos <- svd(w)$u[,1]
w_cols <- rainbow(64)[ceiling(64*(w_pos - min(w_pos))/diff(range(w_pos)))]
layout(matrix(1:ntypes^2, nrow=ntypes), widths=c(1.2,rep(1,ntypes-2),1.1), 
       heights=c(1.5,rep(1,ntypes-2),1.2))
opar <- par("mar")
for (i in 1:ntypes) {
    for (j in 1:ntypes) {
        par(mar=c(opar,0)[c(if (j==ntypes) 1 else 5, 
                            if (i==1) 2 else 5, 
                            if (j==1) 3 else 5, 
                            if (i==ntypes) 4 else 5)])
        plot(w[,i], results$w[,j], col=w_cols, pch=20,
             xaxt=if (j==ntypes) "s" else "n",  
             xlab=if (j==ntypes) "true w" else "",  
             yaxt=if (i==1) "s" else "n",  
             ylab=if (i==1) "inferred w" else "",
             xlim=c(0,1), ylim=c(0,1))
        abline(0,1)
    }
}
```

------------------

```{r hmatch_w, echo=FALSE}
wij <- apply(cor(results$w, w), 2, which.max)
mean_abs_err <- mean(abs(w - results$w[,wij]))
```

The mean absolute error in inference of mixture proportions ($w$)
is `r mean_abs_err`.


## Can we infer expression profiles?

```{r hx_results}
cor(results$x, type_mu)
```

----------------


```{r hx_plot_res, fig.width=4*fig.dim}
xij <- apply(cor(results$x, type_mu), 2, which.max)
x_ord <- apply(type_mu, 2, order)
layout(t(1:ntypes))
for (i in 1:ntypes) {
    plot(type_mu[x_ord[,i],i], pch=20,
         main=paste("expression profile", i))
    points((results$eta * results$x[,xij[i]])[x_ord[,i]], pch=20, col='red')
}
legend("topleft", pch=20, col=c("black", "red"),
       legend=c("truth", "inferred"))
```


